sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala

```diff
@@ -64,10 +65,32 @@ case class BroadcastHashJoinExec(
     val numOutputRows = longMetric("numOutputRows")
 
     val broadcastRelation = buildPlan.executeBroadcast[HashedRelation]()
-    streamedPlan.execute().mapPartitions { streamedIter =>
-      val hashed = broadcastRelation.value.asReadOnlyCopy()
-      TaskContext.get().taskMetrics().incPeakExecutionMemory(hashed.estimatedSize)
-      join(streamedIter, hashed, numOutputRows)
+    if (isNullAwareAntiJoin) {
+      streamedPlan.execute().mapPartitionsInternal { streamedIter =>
+        if (broadcastRelation.value.inputEmpty) {
+          streamedIter
+        } else if (broadcastRelation.value.anyNullKeyExists) {
```

Comment:
I'm a bit confused. If the build side has at least one null key, this join returns no result?

Reply:
Yes. it's what defined in null aware anti join.
you can refer to test case . 
sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/not-in-unit-tests-single-column.sql

![image](https://user-images.githubusercontent.com/17242071/87950345-c16bc180-cad9-11ea-87ae-a3b79cd40f9b.png)

Reply:
Interesting. Then I don't think we need to broadcast at all if there is null key in the build side.

Reply:
and this is why it planed into BroadcastNestedLoopJoinExec, it demand M*N calculation to 
eval condition Or(EqualTo(a=b), IsNull(EqualTo(a=b))

Reply:
> Interesting. Then I don't think we need to broadcast at all if there is null key in the build side (we know it at the driver side).

in the very beginning, i am using a scala HashSet to store these singleColumn data, but committer suggest that I should take advantage of HashedRelation API, so after that, we are moving on to HashedRelation.

Reply:
I mean, when we are building the hash relation at the driver side, if we see null key, we can stop earlier, skip the broadcast step, and just return Nil.

Reply:
We can also expand this optimization to shuffle hash join as well (we don't have to include it in this PR).

Reply:
As for my option, I think we should not use BroadcastHashJoinExec to deal with this special case, because its specialization break the hypothesis of normal join.

normally join case is consist of 
joinKeys and condition

but Or(EqualTo(a=b), IsNull(EqualTo(a=b)) can't trans into such pattern.

Reply:
Yeah, NAAJ is super super crazy and I think @leanken's logic is correct. It cannot be generalized easily to the "distributed" case of shuffled hash join, so let's ONLY fix this for broadcast joins -- This also maintains parity with existing broadcast nested loop join that will invariably do a broadcast only join for NAAJ. (So it's not like we have increased the memory needed by this PR)

Reply:
> We can also expand this optimization to shuffle hash join as well (we don't have to include it in this PR).

Sure. 
I can do some research to see if i can support shuffled hash join and multi column support in next PR, maybe.